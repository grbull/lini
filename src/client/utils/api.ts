import Axios from 'axios';

import { AuthLoginDto, AuthValidateDto } from '../../server/auth/auth.dto';
import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { ShowDto, ShowWithEpisodesDto } from '../../server/show/show.dto';
import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
  // SubscriptionUpdateDto,
} from '../../server/subscription/subscription.dto';
import { UserDto, UserUpdateDto } from '../../server/user/user.dto';

const axios = Axios.create({
  baseURL: process.env.SERVER_URL,
  withCredentials: true,
});

async function userLogin(loginDto: AuthLoginDto): Promise<void> {
  await axios.post<void>('auth/login', loginDto);
}

async function userValidate(validateDto: AuthValidateDto): Promise<UserDto> {
  const response = await axios.post<UserDto>('auth/validate', validateDto);
  return response.data;
}

async function userGet(): Promise<UserDto> {
  const response = await axios.get<UserDto>('user');
  return response.data;
}

async function userUpdate(updateDto: UserUpdateDto): Promise<UserDto> {
  const response = await axios.put<UserDto>('user', updateDto);
  return response.data;
}

async function showSearch(query: string): Promise<ShowDto[]> {
  const response = await axios.get<ShowDto[]>(`show/search/${query}`);
  return response.data;
}

async function showGet(id: number): Promise<ShowWithEpisodesDto> {
  const response = await axios.get<ShowWithEpisodesDto>(`show/${id}`);
  return response.data;
}

async function episodeGet(id: number): Promise<EpisodeDto> {
  const response = await axios.get<EpisodeDto>(`episode/${id}`);
  return response.data;
}

async function episodeGetByShow(id: number): Promise<EpisodeDto[]> {
  const response = await axios.get<EpisodeDto[]>(`episode/show/${id}`);
  return response.data;
}

async function subscriptionGetAll(): Promise<SubscriptionDto[]> {
  const response = await axios.get<SubscriptionDto[]>('subscription/all');
  return response.data;
}

async function subscriptionCreate(
  createDto: SubscriptionCreateDto
): Promise<SubscriptionDto> {
  const response = await axios.post<SubscriptionDto>('subscription', createDto);
  return response.data;
}

// async function subscriptionUpdate(
//   updateDto: SubscriptionUpdateDto
// ): Promise<SubscriptionDto> {
//   const response = await axios.put<SubscriptionDto>('subscription', updateDto);
//   return response.data;
// }

async function subscriptionRemove(
  removeDto: SubscriptionRemoveDto
): Promise<SubscriptionDto> {
  const response = await axios.delete<SubscriptionDto>('subscription', {
    data: removeDto,
  });
  return response.data;
}

async function episodeScheduleGet(): Promise<EpisodeScheduleDto> {
  const response = await axios.get<EpisodeScheduleDto>('episode/schedule');
  return response.data;
}

async function pushCreate(createDto: PushSubscriptionCreateDto): Promise<void> {
  await axios.post('push_subscription', createDto);
}

export const api = {
  user: {
    login: userLogin,
    validate: userValidate,
    get: userGet,
    update: userUpdate,
  },
  show: {
    search: showSearch,
    get: showGet,
  },
  episode: {
    get: episodeGet,
    getByShow: episodeGetByShow,
    getSchedule: episodeScheduleGet,
  },
  subscription: {
    getAll: subscriptionGetAll,
    create: subscriptionCreate,
    // update: subscriptionUpdate,
    remove: subscriptionRemove,
  },
  push: {
    create: pushCreate,
  },
};
