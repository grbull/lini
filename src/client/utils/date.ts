import { DateTime } from 'luxon';

export function dateFromNow(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return DateTime.fromISO(airstamp, { locale }).toRelative() || airstamp;
}

export function dateToLocaleDate(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(DateTime.DATE_FULL) ||
    airstamp
  );
}

export function dateToLocaleTime(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(
      DateTime.TIME_WITH_SHORT_OFFSET
    ) || ''
  );
}
