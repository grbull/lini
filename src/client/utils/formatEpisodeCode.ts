export function formatEpisodeCode(season: number, episode: number): string {
  let s: string = season.toString(10);
  let e: string = episode.toString(10);
  if (season < 10) {
    s = `0${s}`;
  }
  if (episode < 10) {
    e = `0${e}`;
  }
  return `S${s}E${e}`;
}
