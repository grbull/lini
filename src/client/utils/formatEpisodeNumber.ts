export function formatEpisodeNumber(number: number): string {
  if (number < 10) {
    return `0${number.toString(10)}`;
  }
  return number.toString();
}
