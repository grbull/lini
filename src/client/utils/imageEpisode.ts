// TODO: Create placeholder
const placeHolder = '/placeholder_episode.png';

export function imageEpisode(url: string | null): string {
  if (url) {
    return process.env.IMAGE_PROXY_URL + url;
  }
  return placeHolder;
}
