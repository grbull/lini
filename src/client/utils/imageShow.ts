// TODO: Create placeholder
const placeHolder = '/placeholder_show.png';

export function imageShow(url?: string): string {
  if (url) {
    return process.env.IMAGE_PROXY_URL + url;
  }
  return placeHolder;
}
