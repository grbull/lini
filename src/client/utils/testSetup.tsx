import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';

export interface TestSetupProps {
  history?: MemoryHistory;
  state?: Partial<RootState>;
}

export interface TestSetupResult extends RenderResult {
  dispatch: typeof jest.fn;
  history: MemoryHistory<unknown>;
  scrollTo: typeof jest.fn;
}

export function testSetup(
  children: ReactNode,
  props?: TestSetupProps
): TestSetupResult {
  const history = props?.history || createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: props?.state || undefined,
  });

  // const reduxHistory = [];
  // const dispatch = (action: any): any => reduxHistory.push(action);
  const dispatch = jest.fn();
  store.dispatch = dispatch;

  const scrollTo = jest.fn();
  window.scrollTo = scrollTo;

  const utils = render(
    <Router history={history}>
      <Provider store={store}>{children}</Provider>
    </Router>
  );

  return { dispatch, history, scrollTo, ...utils };
}
