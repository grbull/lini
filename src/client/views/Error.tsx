import cn from 'classnames';
import React, { ReactElement } from 'react';

import { ErrorMessage } from '../components/ErrorMessage';
import { PageHeader } from '../components/PageHeader';

interface Props {
  error?: string;
}

export function Error({
  error = 'Something went wrong.',
}: Props): ReactElement {
  return (
    <>
      <PageHeader title="Error" />
      <ErrorMessage className={cn('p-2.5')} illustration>
        {error}
      </ErrorMessage>
    </>
  );
}

Error.defaultProps = { error: 'Something went wrong.' };
