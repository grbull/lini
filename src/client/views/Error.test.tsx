/**
 * @jest-environment jsdom
 */

import React from 'react';

import { testSetup } from '../utils/testSetup';
import { Error } from './Error';

describe('Error View', () => {
  it('matches the snapshot', () => {
    const { asFragment } = testSetup(<Error />);

    expect(asFragment()).toMatchSnapshot();
  });
});
