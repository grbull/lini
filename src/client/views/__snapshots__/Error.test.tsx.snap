// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error View matches the snapshot 1`] = `
<DocumentFragment>
  <header
    class="w-full h-12 flex items-center p-2 md:hidden justify-between"
  >
    <button
      class="w-11 float-left"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-left fa-w-14 m-1 text-lg"
        data-icon="arrow-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mx-2 font-medium text-lg overflow-x-auto whitespace-nowrap"
    >
      Error
    </h1>
    <div
      class="w-11"
    />
  </header>
  <div
    class="px-2 py-2.5 bg-red-400 bg-opacity-50 border border-solid border-red-400 p-2.5"
  >
    <div
      class="flex flex-col items-center justify-center"
    >
      <img
        alt="Frustrated Illustration"
        class="opacity-80 mb-2.5"
        src="/frustrated.png"
        width="80"
      />
    </div>
    <h2
      class="font-bold"
    >
      Error
    </h2>
    <p>
      Something went wrong.
    </p>
  </div>
</DocumentFragment>
`;
